beacon-chain/db/filesystem/iteration.go

```diff
@@ -0,0 +1,230 @@
+
+var errIdentFailure = errors.New("failed to determine blob metadata, ignoring all sub-path.")
+
+type identificationError struct {
+	err   error
+	path  string
+	ident blobIdent
+}
+
+func (ide *identificationError) Error() string {
+	return fmt.Sprintf("%s path=%s, err=%s", errIdentFailure.Error(), ide.path, ide.err.Error())
+}
+
+func (ide *identificationError) Unwrap() error {
+	return ide.err
+}
+
+func (*identificationError) Is(err error) bool {
+	return err == errIdentFailure
+}
+
+func (ide *identificationError) LogFields() logrus.Fields {
+	fields := ide.ident.logFields()
+	fields["path"] = ide.path
+	return fields
+}
+
+func newIdentificationError(path string, ident blobIdent, err error) *identificationError {
+	return &identificationError{path: path, ident: ident, err: err}
+}
+
+func listDir(fs afero.Fs, dir string) ([]string, error) {
+	top, err := fs.Open(dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open directory descriptor")
+	}
+	defer func() {
+		if err := top.Close(); err != nil {
+			log.WithError(err).Errorf("Could not close file %s", dir)
+		}
+	}()
+	// re the -1 param: "If n <= 0, Readdirnames returns all the names from the directory in a single slice"
+	dirs, err := top.Readdirnames(-1)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read directory listing")
+	}
+	return dirs, nil
+}
+
+// identPopulator is a function that sets values in the blobIdent for a given level of the filesystem layout.
+type identPopulator func(blobIdent, string) (blobIdent, error)
+
+type layoutLevel struct {
+	populateIdent identPopulator
+	filter        func(string) bool
+}
+
+// identIterator moves through the filesystem in order to yield blobIdents.
+// layoutLevels (in the 'levels' field) allows a filesystem layout to control how the
+// layout is traversed. A layoutLevel can filter out entries from the directory listing
+// via the filter function, and populate fields in the blobIdent via the populateIdent function.
+// The blobIdent is populated from an empty value at the root, accumulating values for its fields at each level.
+// The fully populated blobIdent is returned when the iterator reaches the leaf level.
+type identIterator struct {
+	fs      afero.Fs
+	path    string
+	child   *identIterator
+	ident   blobIdent
+	levels  []layoutLevel
+	entries []string
+	offset  int
+	eof     bool
+}
+
+// next is the only method that a user of the identIterator needs to call.
+// identIterator will blobIdents in a breadth-first fashion,
+// returning an empty blobIdent and io.EOF once all branches have been traversed.
+func (iter *identIterator) next() (blobIdent, error) {
+	if iter.eof {
+		return blobIdent{}, io.EOF
+	}
+	if iter.child != nil {
+		next, err := iter.child.next()
+		if err == nil {
+			return next, nil
+		}
+		if err != io.EOF {
+			return blobIdent{}, err
+		}
+	}
+	return iter.advanceChild()
+}
+
+// advanceChild is used to move to the next directory at each level of the tree, either when
```

Comment:
I feel confused about the words level and tree.
I'm not sure if we are talking about level (and leaves) of the tree represented by the filesystem, or about levels composed of filters and populate (`identIteratoor.levels`)